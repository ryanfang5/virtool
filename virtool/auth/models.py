from sqlalchemy import Column, Enum, String

from virtool.pg.base import Base
from virtool.pg.utils import SQLEnum


class ResourceType(str, SQLEnum):
    """
    Types of resources a permission can apply to (e.g. 'app', 'sample', 'group').

    """

    app = "app"
    group = "group"


class Action(str, SQLEnum):
    """
    Conserved actions that can be performed on a resource and
    are controlled by permissions.
    """

    create = "create"
    update = "update"
    delete = "delete"
    modify = "modify"
    remove = "remove"


class SQLPermission(Base):
    """
    SQL model to store permissions

    """

    __tablename__ = "permissions"

    id = Column(String, primary_key=True)
    name = Column(String)
    description = Column(String)
    resource_type = Column(Enum(ResourceType))
    action = Column(Enum(Action))
