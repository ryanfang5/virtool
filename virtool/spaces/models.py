from sqlalchemy import (
    Column,
    DateTime,
    Integer,
    String,
)

from virtool.pg.base import Base


class SpaceModel(Base):
    __tablename__ = "space"

    id = Column(Integer, primary_key=True)
    name = Column(String, unique=True)
    description = Column(String)
    created_at = Column(DateTime, nullable=False)
    updated_at = Column(DateTime, nullable=False)
    created_by = Column(String, nullable=False)

