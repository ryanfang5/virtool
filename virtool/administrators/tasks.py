from tempfile import TemporaryDirectory
from typing import Dict

from virtool.data.layer import DataLayer
from virtool.tasks.task import BaseTask


class PromoteAdministratorsTask(BaseTask):
    name = "promote_administrators"

    def __init__(
        self, task_id: int, data: DataLayer, context: Dict, temp_dir: TemporaryDirectory
    ):
        super().__init__(task_id, data, context, temp_dir)

        self.steps = [self.promote_administrators]

    async def promote_administrators(self):
        """
        Promote all users that are administrators to the Full administrator role.

        """
        await self.data.administrators.promote_administrators()
