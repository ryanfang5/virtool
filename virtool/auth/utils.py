import sys
from enum import Enum
from logging import getLogger

import openfga_sdk
from aiohttp import ClientConnectorError
from openfga_sdk import (
    CreateStoreRequest,
    WriteAuthorizationModelRequest,
    TypeDefinition,
    Userset,
    OpenFgaApi,
    WriteRequest,
    TupleKey,
    ApiException,
    TupleKeys,
)
from openfga_sdk.api import open_fga_api

from virtool.auth.permissions import (
    AppPermission,
)

logger = getLogger("openfga")


async def connect_openfga(openfga_host: str, openfga_scheme: str):
    """
    Connects to an OpenFGA server and configures the store id.
    Returns the application client instance.
    """
    configuration = openfga_sdk.Configuration(
        api_scheme=openfga_scheme, api_host=openfga_host
    )

    logger.info("Connecting to OpenFGA")

    try:
        api_client = openfga_sdk.ApiClient(configuration)

        api_instance = open_fga_api.OpenFgaApi(api_client)

        configuration.store_id = await get_or_create_store(api_instance)

        await write_auth_model(api_instance)

    except ClientConnectorError:
        logger.critical("Could not connect")
        sys.exit(1)

    return api_instance


async def get_or_create_store(api_instance: OpenFgaApi):
    """
    Get the OpenFGA Store or create one if it does not exist.

    :return: the store id
    """
    response = await api_instance.list_stores()

    logger.info("Connected to OpenFGA")

    if response.stores:
        logger.info("Found existing store")
        return response.stores[0].id

    body = CreateStoreRequest(
        name="Virtool",
    )
    logger.info("Creating store")
    response = await api_instance.create_store(body)
    return response.id


async def write_auth_model(api_instance: OpenFgaApi):
    """
    Write the authorization model for the OpenFGA Store if it does not exist.
    """
    response = await api_instance.read_authorization_models()

    app_definition = TypeDefinition(
        type="app",
        relations={permission.name: Userset(this={}) for permission in AppPermission},
    )

    group_definition = TypeDefinition(
        type="group",
        relations={"member": Userset(this={})},
    )

    type_definitions = WriteAuthorizationModelRequest(
        type_definitions=[app_definition, group_definition],
    )

    if (
        response.authorization_models
        and response.authorization_models[0].type_definitions
        == type_definitions.type_definitions
    ):
        return

    await api_instance.write_authorization_model(type_definitions)


async def write_tuple(
    api_instance, user_type, user_id, relations, object_type, object_name
):
    """
    Write a relationship tuple in OpenFGA.
    """

    tuple_list = [
        TupleKey(
            user=f"{user_type}:{user_id}",
            relation=relation.name if isinstance(relation, Enum) else relation,
            object=f"{object_type}:{object_name}",
        )
        for relation in relations
    ]

    body = WriteRequest(writes=TupleKeys(tuple_keys=tuple_list))

    try:
        await api_instance.write(body)
    except ApiException:
        pass
