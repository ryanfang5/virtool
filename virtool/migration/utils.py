from pathlib import Path


def get_revisions_path() -> Path:
    """
    Get the hardcoded path to data revision files.

    :return: the revisions path

    """
    return Path(__file__).parent.parent.parent / "assets/revisions"
