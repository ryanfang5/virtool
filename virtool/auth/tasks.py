from typing import Dict, TYPE_CHECKING

from humanfriendly.testing import TemporaryDirectory

from virtool.tasks.task import BaseTask

if TYPE_CHECKING:
    from virtool.data.layer import DataLayer


class SyncPermissionsTask(BaseTask):
    """
    Sync group permissions with OpenFGA.
    """

    name = "sync_permissions"

    def __init__(
        self,
        task_id: int,
        data: "DataLayer",
        context: Dict,
        temp_dir: TemporaryDirectory,
    ):
        super().__init__(task_id, data, context, temp_dir)

        self.steps = [self.sync]

    async def sync(self):
        await self.data.auth.sync()
