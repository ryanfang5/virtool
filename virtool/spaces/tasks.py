from tempfile import TemporaryDirectory
from typing import Dict

from virtool.data.layer import DataLayer
from virtool.tasks.task import BaseTask


class PromoteAdministratorsTask(BaseTask):
    name = "promote_administrators"

    def __init__(
        self, task_id: int, data: DataLayer, context: Dict, temp_dir: TemporaryDirectory
    ):
        super().__init__(task_id, data, context, temp_dir)

        self.steps = [self.promote_administrators]

    async def promote_administrators(self):
        """
        Move existing NuVs analysis files to `<data_path>`/analyses/:id and remove the
        old directory.

        """
        await self.data.administrators.promote_administrators()
