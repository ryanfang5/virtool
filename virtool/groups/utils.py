from __future__ import annotations

from copy import deepcopy
from typing import List, Dict, Tuple

import virtool.users.utils
from virtool.authorization.permissions import SpacePermission
from virtool.authorization.relationships import GroupPermission
from virtool.users.utils import Permission


def merge_group_permissions(groups: List[dict]) -> Dict[str, bool]:
    """
    Return a :class:`dict` of permissions that will be
    inherited by a user belonging to all the passed ``groups``.

    :param groups: a list of group documents.
    :return: a dict keyed by permission names with boolean values
    indicating the state of the permission

    """
    permission_dict = virtool.users.utils.generate_base_permissions()

    for p in Permission:
        for group in groups:
            try:
                if group["permissions"][p.value]:
                    permission_dict[p.value] = True
            except KeyError:
                permission_dict[p.value] = False

    return permission_dict


def adapt_legacy_permissions_dict(permissions_dict: dict[str, bool]):
    adapted = deepcopy(permissions_dict)

    adapted.pop("remove_job", None)

    adapted["create_reference"] = adapted.pop("create_ref")
    adapted["delete_upload"] = adapted.pop("remove_file")
    adapted["create_upload"] = adapted.pop("upload_file")

    return adapted


def convert_permissions_dict_to_relationships(
    group_id: str,
    permissions_dict: dict[str, bool],
) -> Tuple[List[GroupPermission], List[GroupPermission]]:
    adapted_permissions_dict = adapt_legacy_permissions_dict(permissions_dict)

    adds = [
        GroupPermission(group_id, SpacePermission.from_string(permission))
        for permission, value in adapted_permissions_dict.items()
        if value
    ]

    removes = [
        GroupPermission(group_id, SpacePermission.from_string(permission))
        for permission, value in adapted_permissions_dict.items()
        if not value
    ]

    return adds, removes
