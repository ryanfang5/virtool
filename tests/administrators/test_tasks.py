from sqlalchemy.ext.asyncio import AsyncEngine, AsyncSession

from virtool.administrators.tasks import PromoteAdministratorsTask
from virtool.pg.utils import get_row_by_id
from virtool.tasks.models import Task
from virtool.utils import get_temp_dir


async def test_promote_administrators(
    data_layer,
    mongo,
    pg: AsyncEngine,
    snapshot,
    static_time,
    tmpdir,
    fake2,
):

    await fake2.users.create()

    user = await fake2.users.create()

    async with mongo.create_session() as session:
        await mongo.users.update_one(
            {"_id": user.id},
            {"$set": {"administrator": True}},
            session=session,
        )

    async with AsyncSession(pg) as session:
        session.add(
            Task(
                id=1,
                complete=False,
                count=0,
                progress=0,
                context={},
                step="promote_administrators",
                type="promote_administrators",
                created_at=static_time.datetime,
            )
        )
        await session.commit()

    task = PromoteAdministratorsTask(1, data_layer, {}, get_temp_dir())

    await task.run()

    row = await get_row_by_id(pg, Task, 1)
    assert row.complete is True

    assert (
        await data_layer.administrators._authorization_client.list_administrators()
        == [(user.id, "full")]
    )
