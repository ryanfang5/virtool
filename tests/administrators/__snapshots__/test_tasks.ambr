# name: test_promote_administrators[uvloop]
  <class 'dict'> {
    'available_roles': <class 'list'> [
      <class 'dict'> {
        'description': 'Manage who is an administrator and what they can do.',
        'id': <AdministratorRole.FULL: 'full'>,
        'name': 'Full',
      },
      <class 'dict'> {
        'description': 'Manage instance settings and administrative messages.',
        'id': <AdministratorRole.SETTINGS: 'settings'>,
        'name': 'Settings',
      },
      <class 'dict'> {
        'description': 'Manage users in any space. Delete any space.',
        'id': <AdministratorRole.SPACES: 'spaces'>,
        'name': 'Spaces',
      },
      <class 'dict'> {
        'description': 'Create user accounts. Control activation of user accounts.',
        'id': <AdministratorRole.USERS: 'users'>,
        'name': 'Users',
      },
      <class 'dict'> {
        'description': '
          Provides ability to:
          - Create new spaces even if the `Free Spaces` setting is not enabled.
          - Manage HMMs and common references.
          - View all running jobs.
          - Cancel any job.
        ',
        'id': <AdministratorRole.BASE: 'base'>,
        'name': 'Base',
      },
    ],
    'items': <class 'list'> [
      <class 'dict'> {
        'administrator': True,
        'handle': 'zclark',
        'id': 'fb085f7f',
        'role': <AdministratorRole.FULL: 'full'>,
      },
    ],
  }
---
